src/linker/Linker/SerializationMarker.cs

```diff
@@ -0,0 +1,202 @@
+					_trackedTypes = new Dictionary<SerializerKind, HashSet<(TypeDefinition, ICustomAttributeProvider)>> ();
+
+				return _trackedTypes;
+			}
+		}
+
+		HashSet<TypeDefinition> _recursiveTypes;
+		HashSet<TypeDefinition> RecursiveTypes {
+			get {
+				if (_recursiveTypes == null)
+					_recursiveTypes = new HashSet<TypeDefinition> ();
+
+				return _recursiveTypes;
+			}
+		}
+
+		public SerializationMarker (LinkContext context)
+		{
+			_context = context;
+		}
+
+		public bool IsActive (SerializerKind serializerKind) => ActiveSerializers.HasFlag (serializerKind);
+
+		static DependencyKind ToDependencyKind (SerializerKind serializerKind) => serializerKind switch {
+			SerializerKind.DataContractSerializer => DependencyKind.DataContractSerialized,
+			SerializerKind.XmlSerializer => DependencyKind.XmlSerialized,
+			_ => throw new ArgumentException (nameof (SerializerKind))
+		};
+
+		public void TrackForSerialization (TypeDefinition type, ICustomAttributeProvider provider, SerializerKind serializerKind)
+		{
+			if (ActiveSerializers.HasFlag (serializerKind)) {
+				MarkRecursiveMembers (type, new DependencyInfo (ToDependencyKind (serializerKind), provider));
+				return;
+			}
+
+			if (!TrackedTypes.TryGetValue (serializerKind, out var types)) {
+				types = new HashSet<(TypeDefinition, ICustomAttributeProvider)> ();
+				TrackedTypes.Add (serializerKind, types);
+			}
+
+			types.Add ((type, provider));
+		}
+
+		public void Activate (SerializerKind serializerKind)
+		{
+			if (!Enum.IsDefined<SerializerKind> (serializerKind) || serializerKind == SerializerKind.None)
+				throw new ArgumentException ($"Unexpected serializer kind {nameof (serializerKind)}");
+
+			if (ActiveSerializers.HasFlag (serializerKind))
+				return;
+
+			ActiveSerializers |= serializerKind;
+
+			if (!TrackedTypes.TryGetValue (serializerKind, out var types))
+				return;
+
+			foreach (var (type, provider) in types)
+				MarkRecursiveMembers (type, new DependencyInfo (ToDependencyKind (serializerKind), provider));
+
+			TrackedTypes.Remove (serializerKind);
+		}
+
+		public void MarkRecursiveMembers (TypeReference typeRef, in DependencyInfo reason)
+		{
+			MarkRecursiveMembersInternal (typeRef, reason);
+		}
+
+		void MarkRecursiveMembersInternal (TypeReference typeRef, in DependencyInfo reason)
+		{
+			if (typeRef == null)
+				return;
+
+			DependencyInfo typeReason = reason;
+			while (typeRef is GenericInstanceType git) {
+				if (git.HasGenericArguments) {
+					foreach (var argType in git.GenericArguments)
+						MarkRecursiveMembersInternal (argType, new DependencyInfo (DependencyKind.GenericArgumentType, typeRef));
+				}
+				_context.Tracer.AddDirectDependency (typeRef, typeReason, marked: false);
+				typeReason = new DependencyInfo (DependencyKind.ElementType, typeRef);
+				typeRef = (typeRef as TypeSpecification).ElementType;
+			}
+			// This doesn't handle other TypeSpecs. We are only matching what xamarin-android used to do.
```

Comment:
I think we should explicitely handle arrays. The code already does this, because (Array[int]).Resolve() actually resolves to Int32 (and not Array)!. So either own up to this behavior by adding a comment or handle it explicitly in code.

Reply:
The issue is not arrays but other obscure cases like function pointers (that's the one I can think of) - I actually have a test where we can hit the function pointer case, but I'm not sure it's worth handling here. I will update the comment to at least call out how it works for arrays.